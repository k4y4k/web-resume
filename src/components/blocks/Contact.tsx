import * as React from 'react'
import { graphql, useStaticQuery } from 'gatsby'
import tw, { css } from 'twin.macro'
import Address from '../contact/Address'
import Email from '../contact/Email'
import getNetworkUsernames from '../../helpers/getNetworkUsernames'
import GitHub from '../contact/GitHub'
import LinkedIn from '../contact/LinkedIn'
import Twitter from '../contact/Twitter'
import Website from '../contact/Website'

interface PureContactTypes {
  restrictDisplay?: boolean
  email: string
  twitter: string
  github: string
  website: string
  linkedin: string
  address?: string
  postalCode?: string
  city: string
  countryCode: string
  region: string
}

export const ContactStyles = css`
  ${tw`flex flex-wrap text-white place-content-evenly`}
  background-color: #2b0e26;

  a,
  p {
    ${tw`p-4 hover:bg-deeporange transition-all`}
    ${tw`print:p-2`}
  }
`

export const PureContact = ({
  restrictDisplay = true,
  email,
  twitter,
  github,
  website,
  linkedin,
  city,
  region,
  countryCode,
  address,
  postalCode,
}: PureContactTypes): JSX.Element => {
  return (
    <section css={[ContactStyles]} id='contact' data-testid='contact'>
      <Website url={website} />
      <Twitter username={twitter} />
      <GitHub username={github} />
      <LinkedIn username={linkedin} />
      <Email email={email} />
      <Address
        restrictDisplay={restrictDisplay}
        city={city}
        region={region}
        countryCode={countryCode}
        address={address}
        postalCode={postalCode}
      />
    </section>
  )
}

export const Contact = (): JSX.Element => {
  const data = useStaticQuery(graphql`
    {
      file(extension: { eq: "json" }, name: { eq: "data" }) {
        childDataJson {
          basics {
            email
            website
            location {
              city
              region
              countryCode
            }
            profiles {
              network
              username
            }
          }
        }
      }
    }
  `)

  const { email, website } = data?.file.childDataJson.basics

  // extract list of networks
  const { profiles } = data?.file.childDataJson.basics
  const twitter = getNetworkUsernames(profiles, 'twitter')
  const github = getNetworkUsernames(profiles, 'github')
  const linkedin = getNetworkUsernames(profiles, 'linkedin')

  const { city, countryCode, region } = data?.file.childDataJson.basics.location

  const props = {
    email,
    website,
    twitter,
    github,
    linkedin,
    city,
    region,
    countryCode,
  }
  return <PureContact {...props} />
}

export default Contact
