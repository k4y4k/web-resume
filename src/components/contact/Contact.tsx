import * as React from 'react'
import Email from './Email'
import Twitter from './Twitter'
import GitHub from './GitHub'
import Website from './Website'
import LinkedIn from './LinkedIn'
import Phone from './Phone'
import Address from './Address'
import { useStaticQuery, graphql } from 'gatsby'
import getNetworkUsernames from '../../helpers/getNetworkUsernames'

interface PureContactTypes {
  restrictDisplay?: boolean
  email: string
  twitter: string
  github?: string
  website?: string
  linkedin?: string
  phone?: string
  address: string | undefined
  postalCode: string
  city: string
  countryCode: string
  region: string
}

export const PureContact = ({
  restrictDisplay = true,
  email,
  twitter,
  github,
  website,
  linkedin,
  phone,
  address,
  postalCode,
  city,
  countryCode,
  region,
}: PureContactTypes): JSX.Element => {
  return (
    <section id='contact' data-testid='contact'>
      <Email email={email} />
      <Twitter username={twitter} />
      <GitHub username={github} />
      <LinkedIn username={linkedin} />
      <Website url={website} />
      <Phone restrictDisplay={restrictDisplay} num={phone} />
      <Address
        restrictDisplay={restrictDisplay}
        address={address}
        postalCode={postalCode}
        city={city}
        countryCode={countryCode}
        region={region}
      />
    </section>
  )
}

const Contact = (): JSX.Element => {
  const data = useStaticQuery(graphql`
    {
      file(extension: { eq: "json" }, name: { eq: "data" }) {
        childDataJson {
          basics {
            email
            phone
            website
            location {
              address
              city
              countryCode
              postalCode
              region
            }
            profiles {
              network
              username
            }
          }
        }
      }
    }
  `)

  const { phone, email, website } = data?.file.childDataJson.basics

  // extract list of networks
  const { profiles } = data?.file.childDataJson.basics
  const twitter = getNetworkUsernames(profiles, 'twitter')
  const github = getNetworkUsernames(profiles, 'github')
  const linkedin = getNetworkUsernames(profiles, 'linkedin')

  const {
    address,
    city,
    countryCode,
    postalCode,
    region,
  } = data?.file.childDataJson.basics.location

  const props = {
    phone,
    email,
    website,
    twitter,
    github,
    linkedin,
    address,
    city,
    countryCode,
    postalCode,
    region,
  }
  return <PureContact {...props} />
}

export default Contact
