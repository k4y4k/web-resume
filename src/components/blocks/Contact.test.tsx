import * as React from 'react'
import { Contact, PureContact } from './Contact'
import { render, screen, within } from '@testing-library/react'

import { useStaticQuery } from 'gatsby'

describe('<Contact />', () => {
  beforeAll(() =>
    (useStaticQuery as jest.Mock).mockReturnValue({
      file: {
        childDataJson: {
          basics: {
            email: 'john@gmail.com',
            website: 'http://johndoe.com',
            location: {
              city: 'San Francisco',
              region: 'California',
              countryCode: 'US',
            },
            profiles: [
              {
                network: 'Twitter',
                username: 'john',
              },
            ],
          },
        },
      },
    })
  )

  test('handles no data', () => {
    render(
      <PureContact
        address=''
        email=''
        website=''
        twitter=''
        github=''
        linkedin=''
        postalCode=''
        city=''
        countryCode=''
        region=''
      />
    )

    const contactSectionContents = screen.getAllByText('Error', {
      exact: false,
    })

    // github, email, twitter, website, address, linkedin
    expect(contactSectionContents).toHaveLength(6)
  })

  describe('composes all other elements', () => {
    test('email', () => {
      render(
        <PureContact
          email='kayak@example.com'
          address=''
          twitter=''
          website=''
          github=''
          linkedin=''
          postalCode=''
          city=''
          countryCode=''
          region=''
        />
      )

      const contactSection = screen.getByTestId('contact')
      const email = within(contactSection).getByText(/@example.com/gi)
      expect(email).not.toBeFalsy()
    })

    test('twitter', () => {
      render(
        <PureContact
          address=''
          email=''
          github=''
          website=''
          linkedin=''
          postalCode=''
          city=''
          countryCode=''
          region=''
          twitter='kayakSinger1'
        />
      )

      const contactSection = screen.getByTestId('contact')
      const twitter = within(contactSection).getByText(/@kayakSinger1/gi)
      expect(twitter).not.toBeFalsy()
    })

    test('github', () => {
      render(
        <PureContact
          address=''
          email=''
          twitter=''
          linkedin=''
          website=''
          postalCode=''
          city=''
          countryCode=''
          region=''
          github='octocat'
        />
      )

      const contactSection = screen.getByTestId('contact')
      const github = within(contactSection).getByText(/octocat/gi)

      expect(github).not.toBeFalsy()
    })

    test('website', () => {
      render(
        <PureContact
          address=''
          email=''
          twitter=''
          github=''
          linkedin=''
          postalCode=''
          city=''
          countryCode=''
          region=''
          website='example.com'
        />
      )

      const contactSection = screen.getByTestId('contact')
      const website = within(contactSection).getByText(/example.com/)
      expect(website).not.toBeFalsy()
    })

    test('address', () => {
      render(
        <PureContact
          email=''
          twitter=''
          website=''
          github=''
          linkedin=''
          address='2712 Broadway St'
          postalCode='CA 94115'
          city='San Francisco'
          countryCode='US'
          region='California'
        />
      )

      const contactSection = screen.getByTestId('contact')
      const address = within(contactSection).getByText(
        /San Francisco, California/gi
      )
      expect(address).not.toBeFalsy()
    })

    test('linkedin', () => {
      render(
        <PureContact
          address=''
          email=''
          website=''
          twitter=''
          github=''
          postalCode=''
          city=''
          countryCode=''
          region=''
          linkedin='exampledin'
        />
      )

      const contactSection = screen.getByTestId('contact')
      const linkedin = within(contactSection).getByText(/exampledin/)
      expect(linkedin).not.toBeFalsy()
      expect(linkedin).toHaveAttribute(
        'href',
        'https://linkedin.com/in/exampledin'
      )
    })
  })

  test('Hides things appropriately by default', () => {
    render(
      <PureContact
        email='kayak@example.com'
        twitter='kayakSinger1'
        github='octocat'
        website='example.com'
        linkedin='exampledin'
        city='Example City'
        address='2712 Broadway St'
        postalCode='CA 94115'
        countryCode='US'
        region='California'
      />
    )

    const address = screen.getByTestId('address')
    expect(address).toHaveTextContent('Example City, California')
    expect(address).not.toHaveTextContent('2712')
  })

  test('renders OK', () => {
    render(<Contact />)

    expect(screen.getByTestId('contact')).toMatchSnapshot()
  })
})
