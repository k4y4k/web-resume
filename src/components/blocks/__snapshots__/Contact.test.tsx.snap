// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Contact /> renders OK 1`] = `
<section
  css="
  ,display: flex;,flex-wrap: wrap;,--tw-text-opacity: 1;,color: rgba(255, 255, 255, var(--tw-text-opacity));,place-content: space-evenly;,
  background-color: #2b0e26;

  a,
  p {
    ,padding: 1rem;,:hover {,--tw-bg-opacity: 1;,background-color: rgba(181, 61, 39, var(--tw-bg-opacity));,},transition-property: all;,transition-timing-function: cubic-bezier(0.4, 0, 0.2, 1);,transition-duration: 150ms;,
    ,@media print {,padding: 0.5rem;,},
  }
"
  data-testid="contact"
  id="contact"
>
  <a
    href="http://johndoe.com"
    rel="noreferrer"
    target="_blank"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="2"
        x2="22"
        y1="12"
        y2="12"
      />
      <path
        d="M12 2a15.3 15.3 0 0 1 4 10 15.3 15.3 0 0 1-4 10 15.3 15.3 0 0 1-4-10 15.3 15.3 0 0 1 4-10z"
      />
    </svg>
     
    johndoe.com
  </a>
  <a
    href="https://twitter.com/john"
    rel="noreferrer"
    target="_blank"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M23 3a10.9 10.9 0 0 1-3.14 1.53 4.48 4.48 0 0 0-7.86 3v1A10.66 10.66 0 0 1 3 4s-4 9 5 13a11.64 11.64 0 0 1-7 2c9 5 20 0 20-11.5a4.5 4.5 0 0 0-.08-.83A7.72 7.72 0 0 0 23 3z"
      />
    </svg>
     
    @john
  </a>
  <a
    href="https://github.com/undefined"
    rel="noreferrer"
    target="_blank"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 19c-5 1.5-5-2.5-7-3m14 6v-3.87a3.37 3.37 0 0 0-.94-2.61c3.14-.35 6.44-1.54 6.44-7A5.44 5.44 0 0 0 20 4.77 5.07 5.07 0 0 0 19.91 1S18.73.65 16 2.48a13.38 13.38 0 0 0-7 0C6.27.65 5.09 1 5.09 1A5.07 5.07 0 0 0 5 4.77a5.44 5.44 0 0 0-1.5 3.78c0 5.42 3.3 6.61 6.44 7A3.37 3.37 0 0 0 9 18.13V22"
      />
    </svg>
     
  </a>
  <a
    href="https://linkedin.com/in/undefined"
    rel="noreferrer"
    target="_blank"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M16 8a6 6 0 0 1 6 6v7h-4v-7a2 2 0 0 0-2-2 2 2 0 0 0-2 2v7h-4v-7a6 6 0 0 1 6-6z"
      />
      <rect
        height="12"
        width="4"
        x="2"
        y="9"
      />
      <circle
        cx="4"
        cy="4"
        r="2"
      />
    </svg>
     
  </a>
  <a
    href="mailto:john@gmail.com"
  >
    <svg
      fill="none"
      height="1em"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
      />
      <polyline
        points="22,6 12,13 2,6"
      />
    </svg>
     
    john@gmail.com
  </a>
  <div
    css="cursor: default;"
    data-testid="address"
    id="address"
  >
    <p>
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z"
        />
        <circle
          cx="12"
          cy="10"
          r="3"
        />
      </svg>
       
      San Francisco
      , 
      California,
       
      US
    </p>
  </div>
</section>
`;
